Added graceful shutdown hook

freimair opened the pull request:
<!-- 
- make yourself familiar with the CONTRIBUTING.md if you have not already (https://github.com/bisq-network/bisq/blob/master/CONTRIBUTING.md)
- make sure you follow our [coding style guidelines][https://github.com/bisq-network/style/issues)
- pick a descriptive title
- provide some meaningful PR description below
- create the PR
- in case you receive a "Change request" and/or a NACK, please react within 30 days. If not, we will close your PR and it can not be up for compensation.
- After addressing the change request, __please re-request a review!__ Otherwise we might miss your PR as we tend to only look at pull requests tagged with a "review required".
-->

Fixes https://github.com/bisq-network/bisq/issues/3884, fixes https://github.com/bisq-network/bisq/issues/2465

Graceful shutdown has only been done in case of an error or when
using the GUI. A regular eg. seednode shutdown is not covered
though.

Now, a `kill -TERM ...` or a `kill -INT` or `[Ctrl+C]` triggers a graceful shutdown procedure.

Tested on a Bisq app and seednode service launch by firing up and then
- `kill -TERM javapid`
- `kill -INT javapid`
- run from terminal followed by hitting `[Ctrl+C]`

freimair committed:
Added graceful shutdown hook

Graceful shutdown has only be done in case of an error or when
using the GUI. A regular eg. seednode shutdown is not covered
though.

Now, SIGTERM and SIGINT triggers a graceful shutdown procedure.

wiz commented:
@ripcurlx @sqrrm is it too late to get this into v1.2.8 ?

ripcurlx commented on core/src/main/java/bisq/core/app/BisqExecutable.java:
```diff
@@ -100,6 +101,12 @@ protected void doExecute() {
+        // exit gracefully on shutdown
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+                    gracefulShutDown(() -> log.info("gracefulShutDown complete"));
+                }, "Monitor Shutdown Hook ")
```
Is there a reason why you labeled it `Monitor Shutdown Hook `?
```suggestion
                }, "BisqExecutable Shutdown Hook")
```

freimair replied:
copy and paste... my bad

wiz commented:
I was expecting to see the "Bisq is shutting down" modal popup, same as cmd + Q to quit, but it did not appear. Is it really doing the graceful shutdown same as cmd + Q ? Also, can you handle SIGINT as well as SIGTERM ?

wiz reviewed (changes requested):
## Expected Result

Last line of log output should be
```
Mar-18 02:28:35.012 [JavaFX Application Thread] INFO  bisq.core.app.BisqExecutable: Graceful shutdown completed. Exiting now. 
```

## Actual Result

Last line of log output is random

ripcurlx reviewed (changes requested):
NACK - https://github.com/bisq-network/bisq/pull/4047#pullrequestreview-377101559

Could you please have a look at the comment by @wiz? Thanks!

freimair commented:
ah yes, I had the hopes that after achieving the same for the monitor back in the days, it would also work here.

2 issues:
- first, I found and fixed the shutdown chaos https://github.com/bisq-network/bisq/issues/2465. I rearranged the shutdown procedure, now at least via the GUI, graceful shutdown works as expected (to be refined and committed)
- second, the OS does forward the TERM to all child processes and thus, tor gets shut down before we are ready for it. I had solved that somehow for the monitor, but right now, I am having a hard time finding out why it does work there but not here.

stand by...

@wiz: even if it works someday this will never display something in the GUI. The gui does not have the necessary infrastructure.

freimair committed:
Fix shutdown order

freimair committed:
Wait for connections to be closed

The close connection process did fire up worker threads to actually
close the connections. Yet, once all threads have been spawned,
the code proceeds assuming that there are no connections left open
without checking.
This lead to situations where tor has been shutdown already but
open connections. These connections tried to gracefully close but
without tor, that only caused a wall of exceptions.

freimair committed:
Make sure to shut down tor

Here, the tor object is a member variable and there are cases where
this member variable is not set yet.
Situation arose where a sigterm/sigint shutdown is requested and due
to the member variable not set tor was left running.

freimair committed:
Harmonize System::exit usage

The actual System::exit commands have been scattered around various
places in the code. Sometimes, actual system exit depended on the
calling code to reach its end of execution.

freimair committed:
Fix closing the app by gui exit

Before, the graceful shutdown procedures have been executed in the
user thread. However, the sync mechanics for connections/offers
caused a lockup, as some little parts of the code do execute on the
user thread as well.

freimair committed:
Log output shows graceful shutdown message

freimair committed:
Cleanup logging

freimair committed:
Only send message bundle if not stopped

ripcurlx commented on core/src/main/java/bisq/core/offer/OpenOfferManager.java:
```diff
@@ -221,8 +224,25 @@ public void shutDown(@Nullable Runnable completeHandler) {
+
             if (completeHandler != null)
-                UserThread.runAfter(completeHandler, size * 200 + 500, TimeUnit.MILLISECONDS);
+                UserThread.runAfter(completeHandler, 10, TimeUnit.MILLISECONDS);
```
Is there a specific reason why you used 10 milliseconds before triggering the complete handler?

freimair replied:
In fact after digging into this again, I believe it is best to actually revert this change. We do not have any means of checking whether offers have been taken down already. Syncing that up would mean that we need to alter a whole lot of "stopping" business logic in the code. Although, exactly that is due someday, it is imho not necessary now.

ripcurlx commented on p2p/src/main/java/bisq/network/p2p/network/TorNetworkNode.java:
```diff
@@ -187,6 +187,7 @@ private BooleanProperty torNetworkNodeShutDown() {
                 long ts = System.currentTimeMillis();
                 log.debug("Shutdown torNetworkNode");
                 try {
+                    if (tor == null) tor = Tor.getDefault();
```
NIT: I think we decided at some point not to use conditionals within one line anymore

ripcurlx commented on p2p/src/main/java/bisq/network/p2p/network/NetworkNode.java:
```diff
@@ -225,7 +225,7 @@ public void onSuccess(Connection connection) {
 
                 public void onFailure(@NotNull Throwable throwable) {
-                    log.info("onFailure at sendMessage: peersNodeAddress={}\n\tmessage={}\n\tthrowable={}", peersNodeAddress, networkEnvelope.getClass().getSimpleName(), throwable.toString());
+                    log.debug("onFailure at sendMessage: peersNodeAddress={}\n\tmessage={}\n\tthrowable={}", peersNodeAddress, networkEnvelope.getClass().getSimpleName(), throwable.toString());
```
Is there a specific reason why we don't want to have this in our logs anymore?

freimair replied:
this logline is only triggered in a very limited and rare set of situations. One of them being after tor has been shutdown already (but not if tor is gone without a trigger from Bisq). Plus, this log line does not help support. Iff the info is necessary, `--logLevel=DEBUG` is there.

Generally, log outputs are scattered throughout the p2p package. I am currently working on a battle plan to refactor the message handling in order to get more reliability and control over what happens (we do have a lost message every now and then, and I found hints on why that is).
